gpii/node_modules/settingsHandlers/src/RemoteFileSettingsHandler.js

```diff
@@ -0,0 +1,527 @@
+ * @return {Promise} Resolves when complete, with a string value of the stashed file.
+ */
+gpii.settingsHandlers.remoteFileSettingsHandler.applyFile = function (that, targetPath, url, value) {
+    that.path = targetPath;
+    that.value = value;
+    that.url = fluid.stringTemplate(url, {
+        value: encodeURIComponent(that.value)
+    });
+
+    // Generate the download destination file based on the url
+    var sha1 = crypto.createHash("sha1");
+    sha1.update(that.url);
+    that.downloadPath = path.join(that.gpiiSettingsDir, "download", sha1.digest("hex"));
+
+    var stashFile = path.basename(that.path) + "." + (new Date().toISOString().replace(/:/g, "")) + ".gpii-stashed";
+    that.stashPath = path.join(that.gpiiSettingsDir, "stash", stashFile);
+
+    var existed = fs.existsSync(that.path);
+
+    remoteFileSettingsHandler.customInvokes(that, "apply");
+
+    var promise = fluid.promise();
+    fluid.promise.fireTransformEvent(that.events.applyFile).then(function (value) {
+        fluid.log("remoteFileSettingsHandler applied " + that.path + " from " + that.url);
+        promise.resolve(value);
+    }, function (reason) {
+        fluid.log("remoteFileSettingsHandler failed " + that.path + " from " + that.url, ": ", reason);
+        var restorePromise;
+        // try to put things back if it fails.
+        if (fs.existsSync(that.stashPath) === existed) {
+            // Copy the stashed file back to the original place
+            restorePromise = remoteFileSettingsHandler.transferFile(that.stashPath, that.path);
+        } else {
+            restorePromise = fluid.promise().resolve();
+        }
+
+        restorePromise.then(function () {
+            promise.reject(reason);
+        }, function (innerReason) {
+            fluid.log("remoteFileSettingsHandler failed to restore: ", innerReason);
+            promise.reject(reason);
+        });
+    });
+
+    return promise;
+};
+
+
+/**
+ * Restores a file.
+ *
+ * @param {Component} that The gpii.settingsHandlers.remoteFileDownload instance.
+ * @param {String} targetPath The path to the setting file.
+ * @param {String} stashPath The path to the original copy of the setting file.
+ * @param {String} value The setting value.
+ * @return {Promise} Resolves when complete.
+ */
+gpii.settingsHandlers.remoteFileSettingsHandler.restoreFile = function (that, targetPath, stashPath, value) {
+    that.path = targetPath;
+    that.stashPath = stashPath;
+    that.value = value;
+    remoteFileSettingsHandler.customInvokes(that, "restore");
+    return fluid.promise.fireTransformEvent(that.events.restoreFile);
+};
+
+/**
+ * Setup the custom invokers, invokeBefore and invokeAfter.
+ * @param {Component} that The gpii.settingsHandlers.remoteFileDownload instance.
+ * @param {String} action The action being performed by this instance: "apply" or "restore"
+ */
+gpii.settingsHandlers.remoteFileSettingsHandler.customInvokes = function (that, action) {
```

Comment:
Rather than all this customInvokes gubbins, it looks like it would be far simpler just to fix up the lifecycleManager so that it supports asynchronous actions, and then register that at top level in the solutions registry entry - https://issues.gpii.net/browse/GPII-434 suggests it should just be a one-line fix

Reply:
Yes, you always give my conscience a voice. It's needed because I was treating office as a single application in the SR when really, it's not.

In fact, the change to the lifecycleManager isn't required for this particular solution, so I'll not make it unless you think otherwise.

Reply:
removed them